fix: ignore pyright warnings about incompat pathlib types

davidxia opened the pull request:
Running `inv type-check` shows lots of `"pathlib._local.Path" is not assignable
to "pathlib.Path"` warnings. These seem harmless and caused by either the type
checker's inability to see they're the same class or some upstream pathlib bug.

In any case, this change disables pyright's `reportGeneralTypeIssues` for these
specific lines to make the `inv type-check` output less noisy.

<details>
<summary>Before</summary>

```shell
$ inv type-check
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_clustered_functions.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/container_io_manager.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/execution_context.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_tunnel.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/app.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/client.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/cloud_bucket_mount.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/cls.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/container_process.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/dict.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/environments.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/file_io.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/functions.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/io_streams.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/mount.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/network_file_system.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/object.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/parallel_map.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/partial_function.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/proxy.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/queue.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/runner.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/sandbox.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/secret.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/serving.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/snapshot.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/token_flow.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/volume.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_clustered_functions.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/container_io_manager.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/execution_context.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_tunnel.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/app.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/client.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/cloud_bucket_mount.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/cls.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/container_process.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/dict.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/environments.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/file_io.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/functions.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/io_streams.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/mount.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/network_file_system.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/object.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/parallel_map.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/partial_function.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/proxy.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/queue.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/runner.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/sandbox.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/secret.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/serving.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/snapshot.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/token_flow.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/volume.pyi
29 files reformatted
Success: no issues found in 308 source files
/Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:288:34 - error: Argument of type "Path" cannot be assigned to parameter "local_path" of type "Path" in function "_add_local_dir"
    "pathlib.Path" is not assignable to "pathlib._local.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:288:47 - error: Argument of type "PurePosixPath" cannot be assigned to parameter "remote_path" of type "PurePosixPath" in function "_add_local_dir"
    "pathlib.PurePosixPath" is not assignable to "pathlib._local.PurePosixPath" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:288:74 - error: Argument of type "(source: Path) -> bool" cannot be assigned to parameter "ignore" of type "(Path) -> bool" in function "_add_local_dir"
    Type "(source: Path) -> bool" is not assignable to type "(Path) -> bool"
      Parameter 1: type "Path" is incompatible with type "Path"
        "pathlib._local.Path" is not assignable to "pathlib.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:742:41 - error: Argument of type "str | Path" cannot be assigned to parameter "local_path" of type "str | Path" in function "_from_local_file"
    Type "builtins.str | pathlib.Path" is not assignable to type "builtins.str | pathlib._local.Path"
      Type "Path" is not assignable to type "str | Path"
        "Path" is not assignable to "str"
        "pathlib.Path" is not assignable to "pathlib._local.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:813:39 - error: Argument of type "Path" cannot be assigned to parameter "local_path" of type "Path" in function "_add_local_dir"
    "pathlib.Path" is not assignable to "pathlib._local.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:813:57 - error: Argument of type "PurePosixPath" cannot be assigned to parameter "remote_path" of type "PurePosixPath" in function "_add_local_dir"
    "pathlib.PurePosixPath" is not assignable to "pathlib._local.PurePosixPath" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:813:92 - error: Argument of type "(Path) -> bool" cannot be assigned to parameter "ignore" of type "(Path) -> bool" in function "_add_local_dir"
    Type "(pathlib.Path) -> builtins.bool" is not assignable to type "(pathlib._local.Path) -> builtins.bool"
      Parameter 1: type "Path" is incompatible with type "Path"
        "pathlib._local.Path" is not assignable to "pathlib.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:834:68 - error: Argument of type "str | Path" cannot be assigned to parameter "local_path" of type "str | Path" in function "_from_local_file"
    Type "builtins.str | pathlib.Path" is not assignable to type "builtins.str | pathlib._local.Path"
      Type "Path" is not assignable to type "str | Path"
        "Path" is not assignable to "str"
        "pathlib.Path" is not assignable to "pathlib._local.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:876:69 - error: Argument of type "Sequence[str] | ((Path) -> bool)" cannot be assigned to parameter "ignore" of type "Sequence[str] | ((Path) -> bool) | None" in function "_from_local_python_packages"
    Type "Sequence[str] | ((Path) -> bool)" is not assignable to type "Sequence[str] | ((Path) -> bool) | None"
      Type "(Path) -> bool" is not assignable to type "Sequence[str] | ((Path) -> bool) | None"
        "function" is not assignable to "Sequence[str]"
        Type "(pathlib.Path) -> builtins.bool" is not assignable to type "(pathlib._local.Path) -> builtins.bool"
          Parameter 1: type "Path" is incompatible with type "Path"
            "pathlib._local.Path" is not assignable to "pathlib.Path"
        "function" is not assignable to "None" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:949:17 - error: Argument of type "Path" cannot be assigned to parameter "local_path" of type "Path" in function "_add_local_dir"
    "pathlib.Path" is not assignable to "pathlib._local.Path" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:949:35 - error: Argument of type "PurePosixPath" cannot be assigned to parameter "remote_path" of type "PurePosixPath" in function "_add_local_dir"
    "pathlib.PurePosixPath" is not assignable to "pathlib._local.PurePosixPath" (reportArgumentType)
  /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.py:949:62 - error: Argument of type "(Path) -> bool" cannot be assigned to parameter "ignore" of type "(Path) -> bool" in function "_add_local_dir"
    Type "(pathlib.Path) -> builtins.bool" is not assignable to type "(pathlib._local.Path) -> builtins.bool"
      Parameter 1: type "Path" is incompatible with type "Path"
        "pathlib._local.Path" is not assignable to "pathlib.Path" (reportArgumentType)
12 errors, 0 warnings, 0 informations
```
</details>

<details>
<summary>After</summary>

```shell
$ inv type-check
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_clustered_functions.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/container_io_manager.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/execution_context.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_tunnel.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/app.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/client.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/cloud_bucket_mount.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/cls.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/container_process.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/dict.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/environments.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/file_io.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/functions.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/io_streams.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/mount.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/network_file_system.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/object.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/parallel_map.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/partial_function.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/proxy.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/queue.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/runner.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/sandbox.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/secret.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/serving.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/snapshot.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/token_flow.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/volume.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_clustered_functions.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/container_io_manager.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/execution_context.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/_tunnel.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/app.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/client.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/cloud_bucket_mount.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/cls.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/container_process.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/dict.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/environments.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/file_io.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/functions.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/image.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/io_streams.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/mount.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/network_file_system.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/object.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/parallel_map.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/partial_function.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/proxy.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/queue.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/runner.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/sandbox.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/secret.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/serving.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/snapshot.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/token_flow.pyi
Wrote /Users/dxia/src/github.com/modal-labs/modal-client/modal/volume.pyi
29 files reformatted
Success: no issues found in 308 source files
0 errors, 0 warnings, 0 informations
```
</details>

davidxia commented on modal/image.py:
```diff
@@ -831,11 +831,14 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+        self,
+        *modules: str,
+        copy: bool = False,
+        ignore: Optional[Union[Sequence[str], Callable[[Path], bool]]] = NON_PYTHON_FILES,
```
added `Optional` here since that's what `_from_local_python_packages()`'s `ignore` param seems to expect later.

mwaskom commented:
Hm I don't see these warnings? 
```
$ inv type-check
Removed /home/ec2-user/client/modal/_clustered_functions.pyi
Removed /home/ec2-user/client/modal/_runtime/container_io_manager.pyi
Removed /home/ec2-user/client/modal/_runtime/execution_context.pyi
Removed /home/ec2-user/client/modal/_tunnel.pyi
Removed /home/ec2-user/client/modal/app.pyi
Removed /home/ec2-user/client/modal/client.pyi
Removed /home/ec2-user/client/modal/cloud_bucket_mount.pyi
Removed /home/ec2-user/client/modal/cls.pyi
Removed /home/ec2-user/client/modal/container_process.pyi
Removed /home/ec2-user/client/modal/dict.pyi
Removed /home/ec2-user/client/modal/environments.pyi
Removed /home/ec2-user/client/modal/file_io.pyi
Removed /home/ec2-user/client/modal/functions.pyi
Removed /home/ec2-user/client/modal/image.pyi
Removed /home/ec2-user/client/modal/io_streams.pyi
Removed /home/ec2-user/client/modal/mount.pyi
Removed /home/ec2-user/client/modal/network_file_system.pyi
Removed /home/ec2-user/client/modal/object.pyi
Removed /home/ec2-user/client/modal/parallel_map.pyi
Removed /home/ec2-user/client/modal/partial_function.pyi
Removed /home/ec2-user/client/modal/proxy.pyi
Removed /home/ec2-user/client/modal/queue.pyi
Removed /home/ec2-user/client/modal/runner.pyi
Removed /home/ec2-user/client/modal/sandbox.pyi
Removed /home/ec2-user/client/modal/secret.pyi
Removed /home/ec2-user/client/modal/serving.pyi
Removed /home/ec2-user/client/modal/snapshot.pyi
Removed /home/ec2-user/client/modal/token_flow.pyi
Removed /home/ec2-user/client/modal/volume.pyi
Wrote /home/ec2-user/client/modal/_clustered_functions.pyi
Wrote /home/ec2-user/client/modal/_runtime/container_io_manager.pyi
Wrote /home/ec2-user/client/modal/_runtime/execution_context.pyi
Wrote /home/ec2-user/client/modal/_tunnel.pyi
Wrote /home/ec2-user/client/modal/app.pyi
Wrote /home/ec2-user/client/modal/client.pyi
Wrote /home/ec2-user/client/modal/cloud_bucket_mount.pyi
Wrote /home/ec2-user/client/modal/cls.pyi
Wrote /home/ec2-user/client/modal/container_process.pyi
Wrote /home/ec2-user/client/modal/dict.pyi
Wrote /home/ec2-user/client/modal/environments.pyi
Wrote /home/ec2-user/client/modal/file_io.pyi
Wrote /home/ec2-user/client/modal/functions.pyi
Wrote /home/ec2-user/client/modal/image.pyi
Wrote /home/ec2-user/client/modal/io_streams.pyi
Wrote /home/ec2-user/client/modal/mount.pyi
Wrote /home/ec2-user/client/modal/network_file_system.pyi
Wrote /home/ec2-user/client/modal/object.pyi
Wrote /home/ec2-user/client/modal/parallel_map.pyi
Wrote /home/ec2-user/client/modal/partial_function.pyi
Wrote /home/ec2-user/client/modal/proxy.pyi
Wrote /home/ec2-user/client/modal/queue.pyi
Wrote /home/ec2-user/client/modal/runner.pyi
Wrote /home/ec2-user/client/modal/sandbox.pyi
Wrote /home/ec2-user/client/modal/secret.pyi
Wrote /home/ec2-user/client/modal/serving.pyi
Wrote /home/ec2-user/client/modal/snapshot.pyi
Wrote /home/ec2-user/client/modal/token_flow.pyi
Wrote /home/ec2-user/client/modal/volume.pyi
29 files reformatted
Success: no issues found in 307 source files
0 errors, 0 warnings, 0 informations
```

davidxia commented:
@mwaskom maybe it’s my particular setup. I’m using Python 3.13 in a venv created with uv venv. What’s yours?

mwaskom commented:
Our dev environment uses 3.11. I guess that could be it?

davidxia commented:
> Our dev environment uses 3.11. I guess that could be it?

I think that's it. No errors or warnings with `uv venv` 3.11.12. The [dev env Python version is defined here](https://github.com/modal-labs/modal-client/blob/7823bb79d6ca9c01b0e0a75b5dd700bc5bc00e3f/pyproject.toml#L61)?

mwaskom commented:
> The [dev env Python version is defined here](https://github.com/modal-labs/modal-client/blob/7823bb79d6ca9c01b0e0a75b5dd700bc5bc00e3f/pyproject.toml#L61)?

It's defined in our internal dev setup scripts. I think the version provided there just needs to correspond.

mwaskom commented:
If there are false positives when type checking _against_ Modal on other versions of Python that would be good to know about though!

davidxia commented:
I also got this after I ran `inv type-check` again on this branch on py313. Lmk if you'd like me to open a PR to address any of these.

```shell
$ inv type-check
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_clustered_functions.pyi
Removed /Users/dxia/src/github.com/modal-labs/modal-client/modal/_runtime/container_io_manager.pyi
...
29 files reformatted
modal/_utils/function_utils.py:587: error: "FunctionCreateResponse" has no attribute "server_warnings"  [attr-defined]
modal/_utils/function_utils.py:602: error: "FunctionCreateResponse" has no attribute "server_warnings"  [attr-defined]
modal/_functions.py:777: error: Unexpected keyword argument "experimental_options" for "Function"  [call-arg]
modal/_functions.py:840: error: Unexpected keyword argument "experimental_options" for "FunctionData"  [call-arg]
test/conftest.py:1050: error: Item "Function" of "Function | FunctionData" has no attribute "experimental_options"  [union-attr]
test/conftest.py:1050: error: Item "FunctionData" of "Function | FunctionData" has no attribute "experimental_options"  [union-attr]
test/conftest.py:1054: error: Unexpected keyword argument "server_warnings" for "FunctionCreateResponse"  [call-arg]
Found 7 errors in 3 files (checked 309 source files)
```

mwaskom commented:
I think you need to rerun `inv protoc` maybe as those were all added this morning.

davidxia commented:
> If there are false positives when type checking _against_ Modal on other versions of Python that would be good to know about though!

@mwaskom I think this is the issue with py313. https://github.com/davidxia/debug1
